### PR TITLE
Add  `DatetimeFormatter` and fix Datetime in constraints

### LINKED ISSUE #1203: Cannot apply `Inequality` constraint on demo dataset's datetime columns
### Environment Details
* SDV version: 1.0.0 (in progress) + existing versions (0.18.0)
* Python version: 3.8
* Operating System: Linux (Colab Notebook)

### Error Description
The `student_placements_pii` demo dataset contains `start_date` and `end_date` columns that represent dates. They include some missing values.

If I apply an `Inequality` constraint to designate that `start_date < end_date`, then the synthesizer crashes during `fit`.

### Steps to reproduce
```python
from sdv.datasets.demo import download_demo
from sdv.single_table import GaussianCopulaSynthesizer

data, metadata = download_demo(
    modality='single_table',
    dataset_name='student_placements_pii'
)

start_lessthan_end = {
    'constraint_class': 'Inequality',
    'constraint_parameters': {
      'low_column_name': 'start_date',
      'high_column_name': 'end_date'  
    }   
}

synthesizer = GaussianCopulaSynthesizer(metadata)
synthesizer.add_constraints(constraints=[start_lessthan_end])
synthesizer.fit(data)
```

```
AggregateConstraintsError: 
ufunc 'isnan' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
```

### Notes
Both `start_date` and `end_date` columns are being represented as `object` dtypes in pandas. **This should be OK.**

I also tried casting them myself to see if this helps:
```python
data['start_date'] = pd.to_datetime(data['start_date'])
data['end_date'] = pd.to_datetime(data['end_date'])
```

If I do this, then the `fit` calls works, but there is still an error during `sample`:

```
IntCastingNaNError: Error: Sampling terminated. Partial results are stored in a temporary file: .sample.csv.temp. This file will be overridden the next time you sample. Please rename the file if you wish to save these results.
Cannot convert non-finite values (NA or inf) to integer
```

### PR DESCRIPTION
**Pull request by pvk-developer**
Resolves #1203

### CODE REVIEW

File: sdv/data_processing/data_processor.py [hunk 1]
```diff
@@ -465,6 +465,15 @@ def _fit_numerical_formatters(self, data):
                 )
                 self.formatters[column_name].learn_format(data[column_name])
 
+    def _fit_datetime_formatters(self, data):
```
Comment (amontanez24) [inline sdv/data_processing/data_processor.py:468]: This method is so similar to `_fit_numerical_formatters` that we could probably combine them into a `_fit_formatters` and just have different if statements and one loop

**Review by amontanez24 — APPROVED** [review]
One minor comment but otherwise I think this is good to go!

File: sdv/data_processing/datetime_formatter.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class DatetimeFormatter:
+    """Formatter for datetime data.
+
+    Args:
+        datetime_format (str):
+            The strftime to use for parsing time. For more information, see
+            https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior.
+            If ``None`` it will attempt to learn it by itself.
```
Comment (fealho) [inline sdv/data_processing/datetime_formatter.py:14]: Add that it defaults to None.

File: sdv/data_processing/datetime_formatter.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    def format_data(self, column):
+        """Format a column according to the learned format.
+
+        Args:
+            column (pd.Series):
+                Data to format.
+
+        Returns:
+            numpy.ndarray containing the formatted data.
```
Comment (fealho) [inline sdv/data_processing/datetime_formatter.py:39]: I think we usually format it like this:
```
numpy.ndarray:
   Contains the formatted data.
```

**Review by fealho — COMMENTED** [review]
Looks good! I'd only double check which lines you are not testing, because the coverage is going down right now.

File: sdv/data_processing/data_processor.py (hunk 1 above)
Comment (pvk-developer) [reply sdv/data_processing/data_processor.py:468]: I was doubting between that option too. Will address it asap!

**Comment by pvk-developer** [comment]
> Looks good! I'd only double check which lines you are not testing, because the coverage is going down right now.

Thanks for noticing, actually important lines to test since those were the major changes to the `constraints`.

**Pull request merged**